Duration div mul extras

newpavlov opened the pull request:
Successor of #52556.

This PR adds the following `impl`s:
- `impl Mul<Duration> for u32` (to allow `10*SECOND` in addition to `SECOND*10`)
- `impl Mul<f64> for Duration` (to allow `2.5*SECOND` vs `2*SECOND + 500*MILLISECOND`)
- `impl Mul<Duration> for f64`
- `impl MulAssign<f64> for Duration`
- `impl Div<f64> for Duration`
- `impl DivAssign<f64> for Duration`
- `impl Div<Duration> for Duration` (`Output = f64`, can be useful e.g. for `duration/MINUTE`)

`f64` is chosen over `f32` to minimize rounding errors. (52 bits fraction precision vs `Duration`'s ~94 bit)

rust-highfive commented:
r? @withoutboats

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409401553) ([raw log](https://api.travis-ci.org/v3/job/409401553/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:54]    Compiling rustc_msan v0.0.0 (file:///checkout/src/librustc_msan)
[00:03:56] error[E0308]: mismatched types
[00:03:56]    --> libcore/time.rs:535:20
[00:03:56]     |
[00:03:56] 535 |             nanos: nanos as u64,
[00:03:56]     |                    ^^^^^^^^^^^^ expected u32, found u64
[00:03:56] error[E0308]: mismatched types
[00:03:56]    --> libcore/time.rs:561:20
[00:03:56]     |
[00:03:56]     |
[00:03:56] 561 |             nanos: nanos as u64,
[00:03:56]     |                    ^^^^^^^^^^^^ expected u32, found u64
[00:03:56] error[E0308]: mismatched types
[00:03:56]    --> libcore/time.rs:610:20
[00:03:56]     |
[00:03:56]     |
[00:03:56] 610 |             nanos: nanos as u64,
[00:03:56]     |                    ^^^^^^^^^^^^ expected u32, found u64
[00:03:56] error: aborting due to 3 previous errors
[00:03:56] 
[00:03:56] For more information about this error, try `rustc --explain E0308`.
[00:03:56] error: Could not compile `core`.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409406618) ([raw log](https://api.travis-ci.org/v3/job/409406618/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:04:15]    Compiling cc v1.0.18
[00:04:15]    Compiling core v0.0.0 (file:///checkout/src/libcore)
[00:04:15]    Compiling build_helper v0.1.0 (file:///checkout/src/build_helper)
[00:04:15]    Compiling unwind v0.0.0 (file:///checkout/src/libunwind)
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 521 |         if !nanos.is_finite() {
[00:04:18]     |             ^^^^^ help: try: `self.nanos`
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 524 |         if nanos > (u128::MAX as f64) {
[00:04:18]     |            ^^^^^ help: try: `self.nanos`
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 527 |         let nanos_u128 = nanos.round() as u128;
[00:04:18]     |                          ^^^^^ help: try: `self.nanos`
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 547 |         if !nanos.is_finite() {
[00:04:18] 
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 550 |         if nanos > (u128::MAX as f64) {
[00:04:18] 
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 553 |         let nanos_u128 = nanos.round() as u128;
[00:04:18] 
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 596 |         if !nanos.is_finite() {
[00:04:18]     |             ^^^^^ help: try: `self.nanos`
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 599 |         if nanos > (u128::MAX as f64) {
[00:04:18]     |            ^^^^^ help: try: `self.nanos`
[00:04:18] 
[00:04:18] error[E0425]: cannot find value `nanos` in this scope
[00:04:18]     |
[00:04:18]     |
[00:04:18] 602 |         let nanos_u128 = nanos.round() as u128;
[00:04:18]     |                          ^^^^^ help: try: `self.nanos`
[00:04:22]    Compiling compiler_builtins v0.0.0 (file:///checkout/src/rustc/compiler_builtins_shim)
[00:04:22]    Compiling cmake v0.1.31
[00:04:22]    Compiling alloc_jemalloc v0.0.0 (file:///checkout/src/liballoc_jemalloc)
[00:04:26]    Compiling std v0.0.0 (file:///checkout/src/libstd)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409408399) ([raw log](https://api.travis-ci.org/v3/job/409408399/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:04:06]    Compiling rustc_tsan v0.0.0 (file:///checkout/src/librustc_tsan)
[00:04:07]    Compiling rustc_lsan v0.0.0 (file:///checkout/src/librustc_lsan)
[00:04:08]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:04:09]    Compiling rustc_msan v0.0.0 (file:///checkout/src/librustc_msan)
[00:04:11] error[E0599]: no method named `round` found for type `f64` in the current scope
[00:04:11]     |
[00:04:11]     |
[00:04:11] 527 |         let nanos_u128 = nanos_f64.round() as u128;
[00:04:11] 
[00:04:11] 
[00:04:11] error[E0599]: no method named `round` found for type `f64` in the current scope
[00:04:11]     |
[00:04:11]     |
[00:04:11] 553 |         let nanos_u128 = nanos_f64.round() as u128;
[00:04:11] 
[00:04:11] 
[00:04:11] error[E0599]: no method named `round` found for type `f64` in the current scope
[00:04:11]     |
[00:04:11]     |
[00:04:11] 602 |         let nanos_u128 = nanos_f64.round() as u128;
[00:04:11] 
[00:04:11] error: aborting due to 3 previous errors
[00:04:11] 
[00:04:11] For more information about this error, try `rustc --explain E0599`.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409415016) ([raw log](https://api.travis-ci.org/v3/job/409415016/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:55]    Compiling build_helper v0.1.0 (file:///checkout/src/build_helper)
[00:04:01]    Compiling compiler_builtins v0.0.0 (file:///checkout/src/rustc/compiler_builtins_shim)
[00:04:01]    Compiling cmake v0.1.31
[00:04:01]    Compiling alloc_jemalloc v0.0.0 (file:///checkout/src/liballoc_jemalloc)
[00:04:05] error[E0053]: method `div` has an incompatible type for trait
[00:04:05]     |
[00:04:05]     |
[00:04:05] 619 |     fn div(self, rhs: Duration) -> Duration {
[00:04:05]     |                                    ^^^^^^^^ expected f64, found struct `time::Duration`
[00:04:05]    ::: libcore/ops/arith.rs:441:31
[00:04:05]     |
[00:04:05]     |
[00:04:05] 441 |     fn div(self, rhs: RHS) -> Self::Output;
[00:04:05]     |                               ------------ type in trait
[00:04:05]     |
[00:04:05]     = note: expected type `fn(time::Duration, time::Duration) -> f64`
[00:04:05]                found type `fn(time::Duration, time::Duration) -> time::Duration`
[00:04:05]    Compiling std v0.0.0 (file:///checkout/src/libstd)
[00:04:06] error: aborting due to previous error
[00:04:06] 
[00:04:06] For more information about this error, try `rustc --explain E0053`.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409416220) ([raw log](https://api.travis-ci.org/v3/job/409416220/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:04:01]    Compiling rustc_tsan v0.0.0 (file:///checkout/src/librustc_tsan)
[00:04:02]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:04:03]    Compiling rustc_msan v0.0.0 (file:///checkout/src/librustc_msan)
[00:04:04]    Compiling rustc_lsan v0.0.0 (file:///checkout/src/librustc_lsan)
[00:04:05] error[E0277]: cannot divide `f64` by `time::Duration`
[00:04:05]     |
[00:04:05]     |
[00:04:05] 621 |         let nanos1 = (NPS * (self.secs as f64) + (self.nanos as f64)) / rhs;
[00:04:05]     |                                                                       ^ no implementation for `f64 / time::Duration`
[00:04:05]     |
[00:04:05]     = help: the trait `ops::arith::Div<time::Duration>` is not implemented for `f64`
[00:04:05] 
[00:04:05] error[E0277]: cannot divide `f64` by `time::Duration`
[00:04:05]     |
[00:04:05]     |
[00:04:05] 622 |         let nanos2 = (NPS * (self.secs as f64) + (self.nanos as f64)) / rhs;
[00:04:05]     |                                                                       ^ no implementation for `f64 / time::Duration`
[00:04:05]     |
[00:04:05]     = help: the trait `ops::arith::Div<time::Duration>` is not implemented for `f64`
[00:04:05] error: aborting due to 2 previous errors
[00:04:05] 
[00:04:05] For more information about this error, try `rustc --explain E0277`.
[00:04:05] error: Could not compile `core`.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/409418929) ([raw log](https://api.travis-ci.org/v3/job/409418929/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:51]    Compiling unwind v0.0.0 (file:///checkout/src/libunwind)
[00:03:55] warning: unnecessary parentheses around assigned value
[00:03:55]    --> libcore/time.rs:621:22
[00:03:55]     |
[00:03:55] 621 |         let nanos1 = (NPS * (self.secs as f64) + (self.nanos as f64));
[00:03:55]     |
[00:03:55]     = note: #[warn(unused_parens)] on by default
[00:03:55] 
[00:03:55] warning: unnecessary parentheses around assigned value
[00:03:55] warning: unnecessary parentheses around assigned value
[00:03:55]    --> libcore/time.rs:622:22
[00:03:55]     |
[00:03:55] 622 |         let nanos2 = (NPS * (rhs.secs as f64) + (rhs.nanos as f64));
[00:03:55] 
[00:03:57]    Compiling compiler_builtins v0.0.0 (file:///checkout/src/rustc/compiler_builtins_shim)
[00:03:57]    Compiling cmake v0.1.31
[00:03:57]    Compiling alloc_jemalloc v0.0.0 (file:///checkout/src/liballoc_jemalloc)
---
[00:18:35]    Compiling alloc_jemalloc v0.0.0 (file:///checkout/src/liballoc_jemalloc)
[00:18:36] error: unnecessary parentheses around assigned value
[00:18:36]    --> libcore/time.rs:621:22
[00:18:36]     |
[00:18:36] 621 |         let nanos1 = (NPS * (self.secs as f64) + (self.nanos as f64));
[00:18:36]     |
[00:18:36]     = note: `-D unused-parens` implied by `-D warnings`
[00:18:36] 
[00:18:36] error: unnecessary parentheses around assigned value
[00:18:36] error: unnecessary parentheses around assigned value
[00:18:36]    --> libcore/time.rs:622:22
[00:18:36]     |
[00:18:36] 622 |         let nanos2 = (NPS * (rhs.secs as f64) + (rhs.nanos as f64));
[00:18:36] 
[00:18:38]    Compiling std v0.0.0 (file:///checkout/src/libstd)
[00:18:38]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:18:38]    Compiling rustc_msan v0.0.0 (file:///checkout/src/librustc_msan)
---
[00:18:52] Caused by:
[00:18:52]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name core libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=1cbcabaa1ea822b5 -C extra-filename=-1cbcabaa1ea822b5 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps` (exit code: 1)
[00:18:52] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:18:52] expected success, got: exit code: 101
[00:18:52] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1119:9
[00:18:52] travis_fold:end:stage1-std

[00:18:52] travis_time:end:stage1-std:start=1532826884464708842,finish=1532826907795491003,duration=23330782161


[00:18:52] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:18:52] Build completed unsuccessfully in 0:15:02
[00:18:52] Makefile:28: recipe for target 'all' failed
[00:18:52] make: *** [all] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1246c863
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:09fea60c:start=1532826908394679009,finish=1532826908402812828,duration=8133819
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:012d5ebc
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:025e0a61
travis_time:start:025e0a61
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0718980c
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

kennytm commented:
Could you squash all fixups into a single commit?

newpavlov commented:
Done.

kennytm commented on src/libcore/time.rs:
```diff
@@ -501,13 +502,81 @@ impl Mul<u32> for Duration {
+        if !nanos_f64.is_finite() {
+            panic!("got non-finite value when multiplying duration by float");
+        }
+        if nanos_f64 > (u128::MAX as f64) {
```
The upper limit is `u64::MAX * NANOS_PER_SEC` (1.8 × 10<sup>28</sup>) which is a much smaller number than `u128::MAX` (3.4 × 10<sup>38</sup>). If you check for the former you don't need the second `secs > (u64::MAX as u128)` below.

newpavlov replied:
I was conservative about potential edge effects. So you think it will be enough to just check the following condition?
```Rust
nanos_f64 > ((u64::MAX as u128)*(NANOS_PER_SEC as u128)) as f64
```

kennytm commented on src/libcore/time.rs:
```diff
@@ -501,13 +502,81 @@ impl Mul<u32> for Duration {
+
+    fn mul(self, rhs: f64) -> Duration {
+        const NPS: f64 = NANOS_PER_SEC as f64;
+        let nanos_f64 = rhs * (NPS * (self.secs as f64) + (self.nanos as f64));
```
Consider panicking when `!(rhs >= 0)`.

newpavlov replied:
Good point.

scottmcm commented on src/libcore/time.rs:
```diff
@@ -501,13 +502,77 @@ impl Mul<u32> for Duration {
+    fn mul(self, rhs: Duration) -> Duration {
+        const NPS: f64 = NANOS_PER_SEC as f64;
+        if self.is_sign_negative() {
+            panic!("duration can not be multiplied by negative float");
```
This will trigger for a NaN with the sign bit set, which feels like it'd be better reported under non-finite, so maybe just use `self < 0` here?  Also, multiplying a duration by negative zero seems like it should successfully give a zero duration, not a panic.

scottmcm commented on src/libcore/time.rs:
```diff
@@ -501,13 +502,77 @@ impl Mul<u32> for Duration {
+    type Output = Duration;
+
+    fn mul(self, rhs: Duration) -> Duration {
+        rhs.checked_mul(self).expect("overflow when multiplying scalar by duration")
```
Maybe just use `rhs * self` to avoid repeating things?

newpavlov replied:
They have different panic messages "multiplying scalar by duration" vs "multiplying duration by scalar"

newpavlov replied:
If having different panic messages is not important, then I'll change it and `f64` impl to `rhs * self`.

newpavlov commented:
Considering Rust 1.28 release should I change stabilization version to `1.30.0`?

kennytm commented:
@newpavlov  Yes.

TimNN commented:
Ping from triage, @withoutboats  / @rust-lang/libs. This PR requires your review.

TimNN commented:
Ping from triage, @withoutboats / @rust-lang/libs. This PR requires your review (and, I believe an FCP due to the insta-stable impls).

newpavlov commented:
Any news?

alexcrichton commented:
Ah yes, indeed! The libs team discussed this PR yesterday and we concluded that the new implementation for `u32` is fine, but we're a little hesitant to add the `f64` implementations as insta-stable. An alternative that we discussed was to add unstable methods for converting between `f64` and `Duration` which should in theory solve the same crux of the issue here. How does that sound to you?

newpavlov commented:
I think that methods will be obviously less ergonomic compared to the proposed impls. `Duration` situation currently is not ideal (e.g. `Duration::from_millis(n)` is quite annoying to write and noisy to read), which in turn makes code which works with `Duration`s unergonomic (see #51610). It's unfortunate that `impl`s are insta-stable, is there a solution for this on horizon?

Well, I'll replace `impl`s with methods, but honestly I am not happy with such solution. For the reference, what was the main argument against `f64` `impl`s? Is it about the choice between `f32` and `f64` or about having float operations on integer-like type?

alexcrichton commented:
Unfortunately there is no solution on the horizon for unstable impl. It's true that it's less ergonomic than these impls, but that can also be construed as a feature. For example `u32 * f64` is not supported in Rust, you need to do a cast. It was brought up and I personally agreed with it as well that `Duration * f64` is similar that there's some interesting loss of precision going on there which should probably start out as explicit.

If we later determine, though, that this is widely used enough and we need to make it more ergonomic we could reconsider adding such impls

newpavlov commented:
Ah, sorry for the delay. I've updated the code.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/427584950) ([raw log](https://api.travis-ci.org/v3/job/427584950/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:15:10] .................................................................................................... 1800/2164
[01:15:18] .................................................................................................... 1900/2164
[01:15:26] .................................................................................................... 2000/2164
[01:15:37] ..................................................................................i................. 2100/2164
[01:15:41] .....................................i...........FF......F......
[01:15:41] 
[01:15:41] ---- time.rs - time::Duration::div_duration (line 529) stdout ----
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:530:12
[01:15:41]  --> time.rs:530:12
[01:15:41]   |
[01:15:41] 4 | let dur1 = Duration::new(2, 700_000_000);
[01:15:41]   |            ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:531:12
[01:15:41]   |
[01:15:41]   |
[01:15:41] 5 | let dur2 = Duration::new(5, 400_000_000);
[01:15:41]   |            ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] thread 'time.rs - time::Duration::div_duration (line 529)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:15:41] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:15:41] 
[01:15:41] 
[01:15:41] ---- time.rs - time::Duration::div_f64 (line 497) stdout ----
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:498:11
[01:15:41]   |
[01:15:41] 4 | let dur = Duration::new(2, 700_000_000);
[01:15:41]   |           ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:499:31
[01:15:41]   |
[01:15:41]   |
[01:15:41] 5 | assert_eq!(dur.div_f64(3.14), Duration::new(0, 859_872_611));
[01:15:41]   |                               ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:501:33
[01:15:41]   |
[01:15:41]   |
[01:15:41] 7 | assert_eq!(dur.div_f64(3.14e5), Duration::new(0, 8_598));
[01:15:41]   |                                 ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] 
[01:15:41] thread 'time.rs - time::Duration::div_f64 (line 497)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:15:41] 
[01:15:41] ---- time.rs - time::Duration::mul_f64 (line 466) stdout ----
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:467:11
[01:15:41]   |
[01:15:41] 4 | let dur = Duration::new(2, 700_000_000);
[01:15:41]   |           ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:468:31
[01:15:41]   |
[01:15:41]   |
[01:15:41] 5 | assert_eq!(dur.mul_f64(3.14), Duration::new(8, 478_000_000));
[01:15:41]   |                               ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] error[E0433]: failed to resolve. Use of undeclared type or module `Duration`
[01:15:41]  --> time.rs:469:33
[01:15:41]   |
[01:15:41]   |
[01:15:41] 6 | assert_eq!(dur.mul_f64(3.14e5), Duration::new(847_800, 0));
[01:15:41]   |                                 ^^^^^^^^ Use of undeclared type or module `Duration`
[01:15:41] 
[01:15:41] thread 'time.rs - time::Duration::mul_f64 (line 466)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:15:41] 
[01:15:41] failures:
[01:15:41]     time.rs - time::Duration::div_duration (line 529)
[01:15:41]     time.rs - time::Duration::div_f64 (line 497)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

newpavlov commented:
BTW is it correct that we use `use std::time::Duration;` in doctests even though code is in the `libcore`?

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/427634519) ([raw log](https://api.travis-ci.org/v3/job/427634519/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:14:17] .................................................................................................... 1800/2164
[01:14:25] .................................................................................................... 1900/2164
[01:14:34] .................................................................................................... 2000/2164
[01:14:45] ..................................................................................i................. 2100/2164
[01:14:50] .....................................i............F.F....F......
[01:14:50] 
[01:14:50] ---- time.rs - time::Duration::div_duration (line 533) stdout ----
[01:14:50] ---- time.rs - time::Duration::div_duration (line 533) stdout ----
[01:14:50] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:14:50]  --> time.rs:538:17
[01:14:50]   |
[01:14:50] 8 | assert_eq!(dur1.div_duration(dur2), 0.5);
[01:14:50]   |
[01:14:50]   |
[01:14:50]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:14:50] thread 'time.rs - time::Duration::div_duration (line 533)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:14:50] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:14:50] 
[01:14:50] 
[01:14:50] ---- time.rs - time::Duration::div_f64 (line 499) stdout ----
[01:14:50] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:14:50]  --> time.rs:503:16
[01:14:50]   |
[01:14:50] 7 | assert_eq!(dur.div_f64(3.14), Duration::new(0, 859_872_611));
[01:14:50]   |
[01:14:50]   |
[01:14:50]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:14:50] 
[01:14:50] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:14:50]  --> time.rs:505:16
[01:14:50]   |
[01:14:50] 9 | assert_eq!(dur.div_f64(3.14e5), Duration::new(0, 8_598));
[01:14:50]   |
[01:14:50]   |
[01:14:50]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:14:50] 
[01:14:50] thread 'time.rs - time::Duration::div_f64 (line 499)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:14:50] 
[01:14:50] ---- time.rs - time::Duration::mul_f64 (line 466) stdout ----
[01:14:50] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:14:50]  --> time.rs:470:16
[01:14:50]   |
[01:14:50] 7 | assert_eq!(dur.mul_f64(3.14), Duration::new(8, 478_000_000));
[01:14:50]   |
[01:14:50]   |
[01:14:50]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:14:50] 
[01:14:50] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:14:50]  --> time.rs:471:16
[01:14:50]   |
[01:14:50] 8 | assert_eq!(dur.mul_f64(3.14e5), Duration::new(847_800, 0));
[01:14:50]   |
[01:14:50]   |
[01:14:50]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:14:50] 
[01:14:50] thread 'time.rs - time::Duration::mul_f64 (line 466)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:14:50] 
[01:14:50] failures:
[01:14:50]     time.rs - time::Duration::div_duration (line 533)
[01:14:50]     time.rs - time::Duration::div_f64 (line 499)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/427704748) ([raw log](https://api.travis-ci.org/v3/job/427704748/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:12:31] .................................................................................................... 2000/2165
[01:12:41] ...................................................................................i................ 2100/2165
butes to enable
[01:12:46] 
[01:12:46] thread 'time.rs - time::Duration::div_f64 (line 500)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:12:46] 
[01:12:46] ---- time.rs - time::Duration::mul_f64 (line 466) stdout ----
[01:12:46] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:12:46]  --> time.rs:471:16
[01:12:46]   |
[01:12:46] 8 | assert_eq!(dur.mul_f64(3.14), Duration::new(8, 478_000_000));
[01:12:46]   |
[01:12:46]   |
[01:12:46]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:12:46] 
[01:12:46] error[E0658]: use of unstable library feature 'duration_float_ops': duration/floats operations are unstabe
[01:12:46]  --> time.rs:472:16
[01:12:46]   |
[01:12:46] 9 | assert_eq!(dur.mul_f64(3.14e5), Duration::new(847_800, 0));
[01:12:46]   |
[01:12:46]   |
[01:12:46]   = help: add #![feature(duration_float_ops)] to the crate attributes to enable
[01:12:46] 
[01:12:46] thread 'time.rs - time::Duration::mul_f64 (line 466)' panicked at 'couldn't compile the test', librustdoc/test.rs:333:13
[01:12:46] 
[01:12:46] failures:
[01:12:46]     time.rs - time::Duration::div_duration (line 535)
[01:12:46]     time.rs - time::Duration::div_f64 (line 500)
---
travis_time:end:00f602ca:start=1536765984942529025,finish=1536765984947629405,duration=5100380
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0b98863f
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:18397cd0
travis_time:start:18397cd0
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:20387268
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
Thanks for the update! FWIW we were originally thinking of `Duration <-> f64` conversions to avoid the need for a couple inherent methods. Do you think though that'd be too lossy?

newpavlov commented:
I think precision should be the same, though I am not sure if temporarily loosing type information is a good approach. I can add `to_f64` and `from_f64` methods under the same unstable feature and it can be decided later if we need all methods or `Duration <-> f64` conversions should be enough.

scottmcm commented:
C# calls its "`TimeSpan` as floating point seconds" property `TotalSeconds`; I think something like `.total_seconds() -> f64`  and something for the other direction makes sense here.

alexcrichton commented:
@newpavlov ah yeah could the methods be added here? I think it may help clean up the implementation too as it centralizes all the conversion logic, right?

newpavlov commented:
I've added `as_float_secs` and `from_float_secs` methods, I think such naming will be more consistent with existing methods. Also refactored float methods using them, though now I think they'll have one redundant float multiplication (conversion from seconds to nanoseconds, previously code was working directly with nanoseconds), which compiler will not be able to optimize out.

alexcrichton commented on src/libcore/time.rs:
```diff
@@ -458,6 +459,116 @@ impl Duration {
+    #[inline]
+    pub fn mul_f64(self, rhs: f64) -> Duration {
+        let secs = rhs * self.as_float_secs();
+        if !secs.is_finite() {
```
Should these assertions be moved into `float_from_secs` perhaps?

newpavlov replied:
Hm, yes, you are right.

ollie27 commented on src/libcore/time.rs:
```diff
@@ -30,6 +30,7 @@ const NANOS_PER_MILLI: u32 = 1_000_000;
 const NANOS_PER_MICRO: u32 = 1_000;
 const MILLIS_PER_SEC: u64 = 1_000;
 const MICROS_PER_SEC: u64 = 1_000_000;
+const MAX_NANOS_F64: f64 = ((u64::MAX as u128 + 1)*(NANOS_PER_SEC as u128) - 1) as f64;
```
The `- 1` here doesn't change the value because `f64` doesn't have enough precision so this constant ends up being exactly 1ns greater than the largest possible `Duration`. Either this constant needs to be changed to the next smallest `f64` or the check below needs to be changed to `nanos >= MAX_NANOS_F64`. `from_float_secs(2f64.powi(64))` should trigger an overflow but currently it returns a `Duration` of 0ns.

newpavlov replied:
Ah, yes, it should've been `>=`. Fixed.

alexcrichton commented:
@bors: r+

Looks good to me, thanks!

bors commented:
:pushpin: Commit 2aca69757f078ac3aad9c2aff3c357ca35e5c486 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 2aca69757f078ac3aad9c2aff3c357ca35e5c486 -->

alexcrichton commented:
@bors: r-

er actually, mind filing a tracking issue and filling that in?

newpavlov commented:
Done.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit fd7565b076440829b86cc7bc5f2457bf42d43936 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton fd7565b076440829b86cc7bc5f2457bf42d43936 -->

bors commented:
:hourglass: Testing commit fd7565b076440829b86cc7bc5f2457bf42d43936 with merge 3c8c357fcf68c8348b3ced62bc0d01b7d68820f7...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.9039)

kennytm commented:
@bors retry 3 hour timeout
